torchvision/datasets/triplet.py

```diff
@@ -0,0 +1,139 @@
+from .folder import DatasetFolder
+
+from torch.distributions.multinomial import Multinomial
+from torch.distributions.bernoulli import Bernoulli
+
+import torch
+import torch.utils.data as data
+
+import math
+
+
+def get_class_samples(num_classes, samples):
+    """Bins samples with respect to classes
+
+    Args:
+        num_classes (int): number of classes
+        samples (list[tuple]): (sample, class_idx) tuples
+
+    Returns:
+        list[list]: Bins of sample paths, binned by class_idx
+    """
+    class_samples = [[] for _ in range(num_classes)]
+    for sample_path, class_idx in samples:
+        class_samples[class_idx].append(sample_path)
+
+    return class_samples
+
+
+def get_sampled_indices(dist, num_samples):
+    """
+    Samples from distribution till the number of the non-zero entries in the sample
+    equals num_samples
+
+    Args:
+        dist (torch.distributions.distribution.Distribution): A pytorch distribution
+        num_samples (int): number of non-zero samples needed
+
+    Returns:
+        list: indices of the non-zero samples
+    """
+    sample = dist.sample().nonzero().view(-1)
+    while len(sample) != num_samples:
+        sample = dist.sample().nonzero().view(-1)
+
+    return sample.tolist()
+
+
+def generate_triplets(class_samples, num_triplets):
+    """Generates a set of triplets from bins of samples
+
+    Args:
+        class_samples (list[list]): bins of samples, binned by class
+        num_triplets (int): number of triplets to be generated
+
+    Returns:
+        list[tuple]: triplets of the form (anchor, positive, negative)
+    """
+    triplets = []
+    class_dist = Multinomial(2, torch.Tensor([1.] * len(class_samples)))
+    swap = Bernoulli(torch.Tensor([0.5]))
+    for _ in range(num_triplets):
+        pos_cls, neg_cls = get_sampled_indices(class_dist, 2)
+        if swap.sample().item() == 1:
+            pos_cls, neg_cls = neg_cls, pos_cls
+
+        pos_samples, neg_samples = class_samples[pos_cls], class_samples[neg_cls]
+
+        pos_dist = Multinomial(2, torch.Tensor([1.] * len(pos_samples)))
+        neg_dist = Multinomial(1, torch.Tensor([1.] * len(neg_samples)))
+
+        anc_idx, pos_idx = get_sampled_indices(pos_dist, 2)
+        neg_idx = get_sampled_indices(neg_dist, 1)[0]
+        if swap.sample().item() == 1:
+            anc_idx, pos_idx = pos_idx, anc_idx
+
+        triplet = (pos_samples[anc_idx], pos_samples[pos_idx], neg_samples[neg_idx])
+        triplets.append(triplet)
+
+    return triplets
+
+
+class TripletDataset(data.IterableDataset, DatasetFolder):
```

Comment:
I think this should either be an `IterableDataset`, or a `Dataset`, but not both.

Reply:
I thought it was okay to do this since `IterableDataset` extends `Dataset`. But your final comment mitigates this issue by taking in a Dataset instead of a directory path. I will change `TripletDataset` to only extend `IterableDataset`.

Reply:
These changes have been made. `TripletDataset` now only inherits `IterableDataset`.